endpoints/setuid_test.go

```diff
@@ -18,67 +19,163 @@ import (
-	assertIntsMatch(t, http.StatusOK, response.Code)
-	assertHasSyncs(t, response, map[string]string{
-		"pubmatic": "123",
-	})
-}
+func TestSetUIDEndpoint(t *testing.T) {
+	testCases := []struct {
+		uri                   string
+		existingSyncs         map[string]string
+		gdprAllowsHostCookies bool
+		gdprReturnsError      bool
+		expectedSyncs         map[string]string
+		expectedRespMessage   string
+		expectedResponseCode  int
+		description           string
+	}{
+		{
+			uri:                   "/setuid?bidder=pubmatic&uid=123",
+			existingSyncs:         nil,
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         map[string]string{"pubmatic": "123"},
+			expectedResponseCode:  http.StatusOK,
+			description:           "Set uid for valid bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=some-bidder&uid=123",
+			existingSyncs:         nil,
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         nil,
+			expectedResponseCode:  http.StatusBadRequest,
+			description:           "Don't set uid for an unsupported bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=&uid=123",
+			existingSyncs:         nil,
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         nil,
+			expectedResponseCode:  http.StatusBadRequest,
+			description:           "Don't set uid for an empty bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=some-bidder&uid=123",
+			existingSyncs:         map[string]string{"pubmatic": "1234"},
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         nil,
+			expectedResponseCode:  http.StatusBadRequest,
+			description: "No need to set existing syncs back in response for a request " +
+				"to set uid for an unsupported bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=&uid=123",
+			existingSyncs:         map[string]string{"pubmatic": "1234"},
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         nil,
+			expectedResponseCode:  http.StatusBadRequest,
+			description: "No need to set existing syncs back in response for a request " +
+				"to set uid for an empty bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=pubmatic",
+			existingSyncs:         map[string]string{"pubmatic": "1234"},
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         map[string]string{},
+			expectedResponseCode:  http.StatusOK,
+			description:           "Unset uid for a bidder if the request contains an empty uid for that bidder",
+		},
+		{
+			uri:                   "/setuid?bidder=pubmatic&uid=123",
+			existingSyncs:         map[string]string{"rubicon": "def"},
+			gdprAllowsHostCookies: true,
+			expectedSyncs:         map[string]string{"pubmatic": "123", "rubicon": "def"},
+			expectedResponseCode:  http.StatusOK,
+			description:           "Add the uid for the requested bidder to the list of existing syncs",
+		},
+		{
+			uri:                  "/setuid?bidder=pubmatic&uid=123&gdpr=0",
```

Comment:
Do these tests cases rely on the assumption that pubmatic enables GDPR?

Reply:
Well, the tests actually specify if the given host/vendor (in this case `pubmatic`) is allowed under GDPR or not with the `gdprAllowsHostCookies` field in the testCase. We're using the `mockPermsSetUID` implementation of the `Permissions` interface which allows or disallows a vendor based on it's `allowHost` field rather than actually looking at the consent string